include/zrepl_prot.h

```diff
@@ -76,10 +76,12 @@ typedef enum zvol_op_status zvol_op_status_t;
 struct zvol_io_hdr {
 	uint16_t	version;
 	zvol_op_code_t	opcode;
-	zvol_op_status_t status;
 	uint64_t	io_seq;
 	uint64_t	offset;
 	uint64_t	len;
+	uint64_t	checkpointed_io_seq;
```

Comment:
is it an output variable (from replica to target) storing the most recent io number received from target?

Reply:
Yes. Let me explain it a little better way .....

iSCSI controller assigns a monotonically increasing number to each IO which it get from application. Now iSCSI contoller Fan out this IO to multiple replicas with same IO number. 
Now different replicas will take different amount of time to ACK this write. Because of this iSCSI controller need to maintain a queue of pending IOs. "Lowest IO number pending in this queue -1" would be the IO number which has been "make it to disk along with consistency factor"......this IO number would be called as "checkpointed_io_seq" which would be used for rebuild. How this "checkpointed_io_seq" shared across iSCSI controller and among Replicas ?

1. iSCSI controller would send this "Checkpointed_io_seq" to each Replica on each write in header along with current running IO_Seq no.

2. Replica will store this number at an interval of 10 Min to ZAP atttributes. It means we are storing IO_seq no which was seen 10 Min back by all replicas.

3. At time of Handshake with iSCSI controller, this IO_Seq will be read from ZAP attribute and presented to iSCSI controller.

4. iSCSI controller will get this "checkpointed_io_seq" from each replica during handshake and make a decision table. 

5. Now iSCSI controller will decide if replica need rebuild and based on that contoller would start rebuild on that replica.

Reply:
Many thanks for thorough explanation. Makes sense 👍